### PR TITLE
Avoid unnecessary copies of ModelProto from being made in the InferenceSession class

### PR DESCRIPTION
**Pull request by hariharans29**
**Description**: Fixing some instances wherein unnecessary copies were being made of ModelProto instances

**Motivation and Context**
Reduce peak memory usage footprint at model load time

### CODE REVIEW

File: onnxruntime/core/session/inference_session.cc [hunk 1]
```diff
@@ -487,10 +488,10 @@ common::Status InferenceSession::Load(std::istream& model_istream) {
   }
 
   auto loader = [this, &model_istream](std::shared_ptr<onnxruntime::Model>& model) {
-    ModelProto model_proto;
+    auto model_proto = onnxruntime::make_unique<ONNX_NAMESPACE::ModelProto>();
```
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc]: Eventually stored on the heap (https://github.com/microsoft/onnxruntime/blob/c6b499bfc9f92a745b1ab75bc75b698b5b2b9af9/onnxruntime/core/graph/model.cc#L79). Might as well store it on the heap here and pass on the ownership.

**Comment by snnn** [comment]
I tested it with bvlc_alexnet opset8 model from onnx model zoo, it reduced InferenceSession creating time from 790ms to 608ms, so it's about 30% improve.

**Comment by hariharans29** [comment]
> I tested it with bvlc_alexnet opset8 model from onnx model zoo, it reduced InferenceSession creating time from 790ms to 608ms, so it's about 30% improve.

Thanks for the test.

File: onnxruntime/core/graph/model.cc [hunk 2]
```diff
@@ -394,14 +394,14 @@ Status Model::LoadFromBytes(int count, void* p_bytes, /*out*/ ONNX_NAMESPACE::Mo
 
 Status Model::LoadFromBytes(int count, void* p_bytes, /*out*/ std::shared_ptr<Model>& p_model,
                             const IOnnxRuntimeOpSchemaRegistryList* local_registries, const logging::Logger& logger) {
-  ModelProto model_proto;
+  auto model_proto = onnxruntime::make_unique<ModelProto>();
```
Comment (hariharans29) [inline onnxruntime/core/graph/model.cc]: allocate on the heap once, make desired changes to the instance, and pass on ownership downstream when it is not needed anymore.

File: onnxruntime/core/graph/model.cc [hunk 3]
```diff
@@ -441,11 +441,11 @@ Status Model::Load(int fd, ONNX_NAMESPACE::ModelProto& model_proto) {
 
 Status Model::Load(int fd, std::shared_ptr<Model>& p_model, const IOnnxRuntimeOpSchemaRegistryList* local_registries,
                    const logging::Logger& logger) {
-  ModelProto model_proto;
+  auto model_proto = onnxruntime::make_unique<ModelProto>();
```
Comment (hariharans29) [inline onnxruntime/core/graph/model.cc]: allocate on the heap once, make desired changes to the instance, and pass on ownership downstream when it is not needed anymore.

File: onnxruntime/core/session/inference_session.cc [hunk 4]
```diff
@@ -516,21 +518,21 @@ common::Status InferenceSession::Load(const void* model_data, int model_data_len
   }
 
   auto loader = [this, model_data, model_data_len](std::shared_ptr<onnxruntime::Model>& model) {
-    ModelProto model_proto;
+    auto model_proto = onnxruntime::make_unique<ONNX_NAMESPACE::ModelProto>();
```
Comment (hariharans29) [inline onnxruntime/core/session/inference_session.cc]: Same justification as above.

File: onnxruntime/core/graph/model.cc (hunk 2 above)
Comment (pranavsharma) [inline onnxruntime/core/graph/model.cc]: Does the model_proto member inside the Model class have to be heap allocated? It's forcing all call sites to heap allocate as well just like this particular line. ModelProto has a move constructor that can be used to move the local object to the Model class. Doesn't have to be fixed in this PR.

Comment (hariharans29) [reply onnxruntime/core/graph/model.cc]: Thanks. I ll address that in a separate PR.

**Pull request merged**